crates/sui-indexer-alt-graphql/src/api/types/transaction/mod.rs

```diff
@@ -144,6 +159,100 @@ impl Transaction {
             contents,
         }))
     }
+
+    /// Cursor based pagination through transactions based on filters.
+    pub(crate) async fn paginate(
+        ctx: &Context<'_>,
+        scope: Scope,
+        page: Page<CTransaction>,
+        filter: TransactionFilter,
+    ) -> Result<Connection<String, Transaction>, RpcError> {
+        let mut conn = Connection::new(false, false);
+
+        if page.limit() == 0 {
+            return Ok(Connection::new(false, false));
+        }
+
+        let watermarks: &Arc<Watermarks> = ctx.data()?;
+
+        // TODO: Is this a valid way to access the lowest cp available? What is the behavior when we prune while we are paginating?
+        let Some(global_cp_lo) = watermarks
+            .low_watermark()
+            .map(|watermark| watermark.checkpoint())
+        else {
+            return Ok(Connection::new(false, false));
+        };
+
+        let mut tx_digest_keys =
+            match filter.checkpoint_bounds(scope.checkpoint_viewed_at(), global_cp_lo) {
+                // Determine tx_sequence_numbers if we have checkpoint bounds.
+                Some(cp_bounds) => {
+                    Self::tx_sequence_numbers_by_checkpoint(ctx, cp_bounds, &page).await?
+                }
+                None => return Ok(Connection::new(false, false)),
+            };
```

Comment:
nit: I would use `let Some(...) = ... else { }` here, because in future there will be multiple different ways to get this list of sequence numbers.

Also, keep private helper functions as free, top-level functions (not bound to the type). We may also want to move them into their own module, if they get complicated enough.

```suggestion
        let Some(cp_bounds) = filter.checkpoint_bounds(scope.checkpoint_viewed_at, global_cp_lo) else {
            return Ok(Connection::new(false, false));
        };
           
        let mut tx_digest_keys =
            Self::tx_sequence_numbers_by_checkpoint(ctx, cp_bounds, &page).await?;
```